### PR TITLE
feat: Add `cloudavenue_alb_pool`

### PR DESCRIPTION
**Pull request by gaetanars**
### Description of your changes
Add `cloudavenue_alb_pool` resource and datasource

If you submit change in the provider code, please make sure to:

- [x] Write or modify examples in `examples/` directory
- [x] Write or modify acceptance tests
- [x] Run `make generate` to ensure the doc was updated properly

### How has this code been tested
```
❯ TF_ACC=1 go test -v -count=1 ./internal/tests/alb
=== RUN   TestAccAlbPoolDataSource
--- PASS: TestAccAlbPoolDataSource (6.39s)
=== RUN   TestAccAlbPoolResource
--- PASS: TestAccAlbPoolResource (18.08s)
PASS
ok      github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/tests/alb 24.478s
```

### CODE REVIEW

File: internal/provider/alb/pool_common.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+				Enabled:   types.BoolValue(poolMember.Enabled),
+				IPAddress: types.StringValue(poolMember.IpAddress),
+				Port:      types.Int64Value(int64(poolMember.Port)),
+				Ratio:     types.Int64Value(int64(*poolMember.Ratio)),
+			})
+		}
+	}
+
+	return members
+}
```
Comment (azrod) [inline internal/provider/alb/pool_common.go:34]:
```suggestion
func processMembers(poolMembers []govcdtypes.NsxtAlbPoolMember) (members []member) {
	for _, poolMember := range poolMembers {
		members = append(members, member{
			Enabled:   types.BoolValue(poolMember.Enabled),
			IPAddress: types.StringValue(poolMember.IpAddress),
			Port:      types.Int64Value(int64(poolMember.Port)),
			Ratio:     types.Int64Value(int64(*poolMember.Ratio)),
		})
	}
}
```

File: internal/provider/alb/pool_schema.go [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+					},
+				},
+				Resource: &schemaR.StringAttribute{
+					MarkdownDescription: " should be created.",
+					PlanModifiers: []planmodifier.String{
+						stringplanmodifier.RequiresReplace(),
+					},
+				},
+				DataSource: &schemaD.StringAttribute{
+					MarkdownDescription: " was created.",
```
Comment (azrod) [inline internal/provider/alb/pool_schema.go:116]:
```suggestion
					MarkdownDescription: "was created.",
```

File: internal/provider/alb/pool_schema.go [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+			"edge_gateway_name": superschema.StringAttribute{
+				Common: &schemaR.StringAttribute{
+					MarkdownDescription: "Edge gateway Name in which ALB Pool",
+					Optional:            true,
+					Validators: []validator.String{
+						stringvalidator.ExactlyOneOf(path.MatchRoot("edge_gateway_id"), path.MatchRoot("edge_gateway_name")),
+					},
+				},
+				Resource: &schemaR.StringAttribute{
+					MarkdownDescription: " should be created.",
```
Comment (azrod) [inline internal/provider/alb/pool_schema.go:128]:
```suggestion
					MarkdownDescription: "should be created.",
```

File: internal/provider/alb/pool_schema.go [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+					PlanModifiers: []planmodifier.String{
+						stringplanmodifier.RequiresReplace(),
+					},
+				},
+				DataSource: &schemaD.StringAttribute{
+					MarkdownDescription: " was created.",
```
Comment (azrod) [inline internal/provider/alb/pool_schema.go:134]:
```suggestion
					MarkdownDescription: "was created.",
```

File: internal/provider/alb/pool_schema.go [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+							Optional: true,
+							Default:  booldefault.StaticBool(true),
+						},
+					},
+					"ip_address": superschema.StringAttribute{
+						Common: &schemaR.StringAttribute{
+							MarkdownDescription: "IP address of pool member.",
+						},
+						Resource: &schemaR.StringAttribute{
+							Required: true,
```
Comment (azrod) [inline internal/provider/alb/pool_schema.go:217]: Add validator `IsIP`

File: internal/provider/alb/pool_resource.go [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+	}
+
+	// Store ID
+	plan.ID = utils.StringValueOrNull(createdAlbPool.NsxtAlbPool.ID)
+
+	// Set state to fully populated data
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
```
Comment (azrod) [inline internal/provider/alb/pool_resource.go:138]:
```suggestion
```

File: internal/provider/alb/pool_resource.go [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+	// Set members
+	members := processMembers(albPool.NsxtAlbPool.Members)
+
+	if len(members) > 0 {
+		plan.Members, diags = types.SetValueFrom(ctx, types.ObjectType{AttrTypes: memberAttrTypes}, members)
+		resp.Diagnostics.Append(diags...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
```
Comment (azrod) [inline internal/provider/alb/pool_resource.go:197]:
```suggestion
	if members := processMembers(albPool.NsxtAlbPool.Members); len(members) > 0 {
        plan.Members, diags = types.SetValueFrom(ctx, types.ObjectType{AttrTypes: memberAttrTypes}, members)
        resp.Diagnostics.Append(diags...)
        if resp.Diagnostics.HasError() {
            return
        }
    }
```

File: internal/provider/alb/pool_resource.go [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+
+	// Set refreshed state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
```
Comment (azrod) [inline internal/provider/alb/pool_resource.go:225]:
```suggestion
```

File: internal/provider/alb/pool_resource.go [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+	if err != nil {
+		resp.Diagnostics.AddError("Unable to update ALB Pool", err.Error())
+		return
+	}
+
+	// Set state to fully populated data
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
```
Comment (azrod) [inline internal/provider/alb/pool_resource.go:276]:
```suggestion
```

File: internal/provider/alb/pool_resource.go [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+	healthMonitorSlice := make([]govcdtypes.NsxtAlbPoolHealthMonitor, 0)
+
+	for _, healthMonitor := range healthMonitors {
+		healthMonitorSlice = append(healthMonitorSlice, govcdtypes.NsxtAlbPoolHealthMonitor{
+			Type: healthMonitor,
+		})
+	}
+
+	return healthMonitorSlice, nil
+}
```
Comment (azrod) [inline internal/provider/alb/pool_resource.go:468]: Passing `context.Background()` in the functions `As` and `ElementsAs` is working correctly ?

File: internal/provider/common/org/edgegw.go [hunk 11]
```diff
@@ -33,3 +40,45 @@ func (o *Org) GetEdgeGateway(egw edgegw.BaseEdgeGW) (edgegw.EdgeGateway, error)
...
+		return egw.GetID(), nil
+	}
+
+	edgeGw, err := o.GetNsxtEdgeGatewayByName(egw.GetName())
+	if err != nil {
+		return "", err
+	}
+
+	return edgeGw.EdgeGateway.ID, nil
+}
```
Comment (azrod) [inline internal/provider/common/org/edgegw.go:60]: This is unnecessary. The function `GetEdgeGateway` accept `ID` or `Name` and return `edgegw.EdgeGateway`. `edgegw.EdgeGateway` has func `GetID()`

File: internal/provider/alb/pool_resource.go [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+		resp.Diagnostics.AddError(
+			"Unexpected Import Identifier",
+			fmt.Sprintf("Expected import identifier with format: edge_gateway_name.alb_pool_name. Got: %q", req.ID),
+		)
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("edge_gateway_name"), idParts[0])...)
+	resp.Diagnostics.Append(resp.State.SetAttribute(ctx, path.Root("name"), idParts[1])...)
+}
```
[the remark this answers is not in the document]
Comment (gaetanars) [reply internal/provider/alb/pool_resource.go:329]: I'm not convinced by complicated import functions. The simpler is easier, the user can read on UI the two named attributed, it's more complicated to get an ID for the user.

File: internal/provider/alb/pool_common.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+func processPersistenceProfile(poolPersistenceProfile *govcdtypes.NsxtAlbPoolPersistenceProfile) persistenceProfile {
+	if poolPersistenceProfile != nil {
+		return persistenceProfile{}
+	}
+
+	return persistenceProfile{
+		Type:  types.StringValue(poolPersistenceProfile.Type),
+		Value: utils.StringValueOrNull(poolPersistenceProfile.Value),
+	}
+}
```
Comment (dmicheneau) [inline internal/provider/alb/pool_common.go:51]:
```suggestion
func processPersistenceProfile(poolPersistenceProfile *govcdtypes.NsxtAlbPoolPersistenceProfile) persistenceProfile {
	if poolPersistenceProfile == nil {
		return persistenceProfile{}
	}

	return persistenceProfile{
		Type:  types.StringValue(poolPersistenceProfile.Type),
		Value: utils.StringValueOrNull(poolPersistenceProfile.Value),
	}
}
```

File: internal/provider/common/edgegw/edgegw.go [hunk 14]
```diff
@@ -60,3 +65,15 @@ func (e EdgeGateway) GetName() string {
...
+func (e EdgeGateway) Lock(ctx context.Context) (d diag.Diagnostics) {
+	gwMutexKV.KvLock(ctx, e.GetID())
+	return
+}
+
+// Unlock unlocks the Edge Gateway.
+func (e EdgeGateway) Unlock(ctx context.Context) (d diag.Diagnostics) {
+	gwMutexKV.KvUnlock(ctx, e.GetID())
+	return
+}
```
Comment (dmicheneau) [inline internal/provider/common/edgegw/edgegw.go:79]: I think this is always return an empty diag.Diagnostics

File: internal/provider/alb/pool_resource.go [hunk 15]
```diff
@@ -0,0 +1,472 @@
...
+	// Set health monitors.
+	healtMonitors := processHealthMonitors(albPool.NsxtAlbPool.HealthMonitors)
+
+	if len(healtMonitors) > 0 {
+		plan.HealthMonitors, diags = types.SetValueFrom(ctx, types.StringType, healtMonitors)
+		resp.Diagnostics.Append(diags...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
```
Comment (dmicheneau) [inline internal/provider/alb/pool_resource.go:208]:
```suggestion
	if len(healthMonitors) > 0 {
		plan.HealthMonitors, diags = types.SetValueFrom(ctx, types.StringType, healtMonitors)
		resp.Diagnostics.Append(diags...)
		if resp.Diagnostics.HasError() {
			return
		}
	}
```

**Review by dmicheneau — DISMISSED** [review]

File: internal/tests/alb/pool_resource_test.go [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+			// Read testing
+			{
+				// Apply test
+				Config: testAccAlbPoolResourceConfig,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestMatchResourceAttr(resourceName, "id", regexp.MustCompile(`(urn:vcloud:loadBalancerPool:[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})`)),
+					resource.TestCheckResourceAttr(resourceName, "name", "Example"),
+				),
+			},
+			// Uncomment if you want to test update or delete this block
```
Comment (eremid) [inline internal/tests/alb/pool_resource_test.go:62]:
```suggestion
```

File: internal/tests/alb/pool_resource_test.go [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { tests.TestAccPreCheck(t) },
+		ProtoV6ProviderFactories: tests.TestAccProtoV6ProviderFactories,
+		Steps: []resource.TestStep{
+			// Read testing
+			{
+				// Apply test
+				Config: testAccAlbPoolResourceConfig,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestMatchResourceAttr(resourceName, "id", regexp.MustCompile(`(urn:vcloud:loadBalancerPool:[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})`)),
```
Comment (eremid) [inline internal/tests/alb/pool_resource_test.go:58]:
```suggestion
					resource.TestMatchResourceAttr(resourceName, "id", regexp.MustCompile(`urn:vcloud:loadBalancerPool:[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}`)),
```
Capturing is useful ?

File: internal/tests/alb/pool_resource_test.go [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+					resource.TestCheckResourceAttr(resourceName, "name", "Example"),
+				),
+			},
+			// Uncomment if you want to test update or delete this block
+			{
+				// Update test
+				Config: testAccAlbPoolResourceConfigUpdate,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestMatchResourceAttr(resourceName, "id", regexp.MustCompile(`(urn:vcloud:loadBalancerPool:[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})`)),
```
Comment (eremid) [inline internal/tests/alb/pool_resource_test.go:67]:
```suggestion
					resource.TestMatchResourceAttr(resourceName, "id", regexp.MustCompile(`urn:vcloud:loadBalancerPool:[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}`)),
```

**Review by eremid — APPROVED** [review]

File: internal/provider/alb/pool_datasource.go [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+			return nil, fmt.Errorf("could not retrieve Edge gateway '%s'", d.edgegw.GetIDOrName())
+		}
+		albPool, err = d.client.Vmware.GetAlbPoolByName(nsxtEdge.EdgeGateway.ID, d.GetName())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return albPool, err
+}
```
Comment (azrod) [inline internal/provider/alb/pool_datasource.go:178]:
```suggestion
// GetAlbPool returns the govcd.NsxtAlbPool.
func (d *albPoolDataSource) GetAlbPool() (*govcd.NsxtAlbPool, error) {
	if d.GetID() != "" {
		return d.client.Vmware.GetAlbPoolById(d.GetID())
	}

	nsxtEdge, err := d.org.GetEdgeGateway(d.edgegw)
	if err != nil {
		return nil, fmt.Errorf("could not retrieve Edge gateway '%s'", d.edgegw.GetIDOrName())
	}
	return d.client.Vmware.GetAlbPoolByName(nsxtEdge.GetID(), d.GetName())
}
```

File: internal/provider/alb/pool_resource.go [hunk 20]
```diff
@@ -0,0 +1,463 @@
...
+			return nil, fmt.Errorf("could not retrieve Edge gateway '%s'", r.edgegw.GetIDOrName())
+		}
+		albPool, err = r.client.Vmware.GetAlbPoolByName(nsxtEdge.EdgeGateway.ID, r.GetName())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return albPool, err
+}
```
Comment (azrod) [inline internal/provider/alb/pool_resource.go:360]:
```suggestion
func (d *albPoolDataSource) GetAlbPool() (*govcd.NsxtAlbPool, error) {
	if d.GetID() != "" {
		return d.client.Vmware.GetAlbPoolById(d.GetID())
	}

	nsxtEdge, err := d.org.GetEdgeGateway(d.edgegw)
	if err != nil {
		return nil, fmt.Errorf("could not retrieve Edge gateway '%s'", d.edgegw.GetIDOrName())
	}
	return d.client.Vmware.GetAlbPoolByName(nsxtEdge.GetID(), d.GetName())
}
```

File: internal/provider/alb/pool_resource.go [hunk 21]
```diff
@@ -0,0 +1,463 @@
...
+	if diags.HasError() {
+		return nil, errors.New(diags[0].Detail())
+	}
+
+	persistenceProfile := &govcdtypes.NsxtAlbPoolPersistenceProfile{}
+	persistenceProfile.Type = p.Type.ValueString()
+	persistenceProfile.Value = p.Value.ValueString()
+
+	return persistenceProfile, nil
+}
```
Comment (azrod) [inline internal/provider/alb/pool_resource.go:444]:
```suggestion
func (r *albPoolResource) getAlbPoolPersistenceProfileType(ctx context.Context, d *albPoolModel) (*govcdtypes.NsxtAlbPoolPersistenceProfile, error) {
	if d.PersistenceProfile.IsNull() {
		return nil, ErrPersistenceProfileIsEmpty
	}

	p := &persistenceProfile{}
	if diags := d.PersistenceProfile.As(ctx, p, basetypes.ObjectAsOptions{
		UnhandledNullAsEmpty:    false,
		UnhandledUnknownAsEmpty: false,
	}); diags.HasError() {
		return nil, errors.New(diags[0].Detail())
	}

	return &govcdtypes.NsxtAlbPoolPersistenceProfile{
		Type:  p.Type.ValueString(),
		Value: p.Value.ValueString(),
	}, nil
}
```

File: internal/provider/alb/pool_resource.go [hunk 22]
```diff
@@ -0,0 +1,463 @@
...
+	healthMonitorSlice := make([]govcdtypes.NsxtAlbPoolHealthMonitor, 0)
+
+	for _, healthMonitor := range healthMonitors {
+		healthMonitorSlice = append(healthMonitorSlice, govcdtypes.NsxtAlbPoolHealthMonitor{
+			Type: healthMonitor,
+		})
+	}
+
+	return healthMonitorSlice, nil
+}
```
Comment (azrod) [inline internal/provider/alb/pool_resource.go:463]:
```suggestion
func (r *albPoolResource) getAlbPoolHealthMonitorType(ctx context.Context, d *albPoolModel) (healthMonitors []govcdtypes.NsxtAlbPoolHealthMonitor, err error) {
	var healthMonitorsSlice []string
	
	if diags := d.HealthMonitors.ElementsAs(ctx, &healthMonitorsSlice, true); diags.HasError() {
		return nil, errors.New(diags[0].Detail())
	}

	for _, healthMonitor := range healthMonitorsSlice {
		healthMonitors = append(healthMonitors, govcdtypes.NsxtAlbPoolHealthMonitor{
			Type: healthMonitor,
		})
	}

	return
}
```

File: internal/provider/alb/pool_schema.go [hunk 23]
```diff
@@ -0,0 +1,323 @@
...
+var persistenceProfileAttrTypes = map[string]attr.Type{
+	"type":  types.StringType,
+	"value": types.StringType,
+}
+
+/*
+userSchema
+
+This function is used to create the schema for the user resource and datasource.
+*/
```
Comment (azrod) [inline internal/provider/alb/pool_schema.go:74]:
```suggestion
/*
albPoolSchema

This function is used to create the schema for the ALB Pool resource and datasource.
*/
```

File: internal/provider/alb/pool_schema.go [hunk 24]
```diff
@@ -0,0 +1,323 @@
...
+			},
+			"default_port": superschema.Int64Attribute{
+				Common: &schemaR.Int64Attribute{
+					MarkdownDescription: "Default Port defines destination server port used by the traffic sent to the member.",
+					Computed:            true,
+				},
+				Resource: &schemaR.Int64Attribute{
+					Optional: true,
+					Default:  int64default.StaticInt64(80),
+				},
```
Comment (azrod) [inline internal/provider/alb/pool_schema.go:179]: Add `int64validator.Between(1, 65535),`

**Review by azrod — APPROVED** [review]

**Pull request merged**